### DISCUSSION TITLE
Keyboard accessibility: Improve shortcuts and navigation patterns

### DISCUSSION THREAD

**Discussion by AmeliaBR** [NickColley/semaphore#30] [New features and ideas]
The existing keyboard handling in Pinafore is a *huge* improvement on the default Mastodon web apps, but it does copy over some problematic patterns. 

The biggest problem, to me, is that it overrides the meaning of the up/down arrow keys (which usually allow fine-tuning scroll position) to duplicate the next/previous post behavior of the j/k keys. (See https://github.com/nolanlawson/pinafore/issues/2171 for someone else confused by the jumping behavior.) In the case of long posts and short screens, this makes the end of each post inaccessible without a non-keyboard way to scroll.

Another issue I've noticed: The pop-up dialog displaying the keyboard shortcuts can't be scrolled via keyboard when it doesn't fit on screen!

Other things to discuss:

- Can keyboard shortcuts be user-customizable? This is very helpful for someone using a non-Latin keyboard layout, so they don't have to switch keyboards between browsing and writing. More generally, non-English users might not remember all the English mnemonics behind letter choices, or someone might want to make it easier to browse one-handed.

- Should there be an option to switch to shortcuts more familiar to Twitter users (or users of other apps)? (e.g. L for like instead of F for favourite; n for a new post; O opens media instead of the conversation, which opens on Enter)

- Could there be a way to display the shortcut characters next to the buttons they trigger, instead of displaying the full dialog list?

- Are there actions missing shortcuts, or other ways to improve the keyboard shortcut system? 

[Aside: I will probably not be monitoring / participating in this discussion, but I appreciate all of you who are spending time building better apps!]

**Comment by NickColley** [comment]
Thanks for this a lot of valid points, will spend some time thinking about it and see if there's improvements I can make.

**Comment by NCW-Srandall** [comment]
As a screen reader user, the up/down movement through toots is very much the expected behaviour for me. Of course I presume a screen reader simply announces the whole post regardless of its visibility, so I understand how people more accustomed to visual keyboard navigation through the web might find the experience different. I'm sure I could adjust to the JK rather than using the arrows, but it would be an adjustment.
